lib/ansible/modules/windows/win_blockinfile.ps1

```diff
@@ -0,0 +1,330 @@
+    $encodingObject = [System.Text.UTF8Encoding] $false;
+
+    # Get a sorted list of encodings with preambles, longest first
+    $max_preamble_len = 0;
+    $sortedlist = New-Object System.Collections.SortedList;
+    Foreach ($encodinginfo in [System.Text.Encoding]::GetEncodings()) {
+        $encoding = $encodinginfo.GetEncoding();
+        $plen = $encoding.GetPreamble().Length;
+        If ($plen -gt $max_preamble_len) {
+            $max_preamble_len = $plen;
+        }
+        If ($plen -gt 0) {
+            $sortedlist.Add( - ($plen * 1000000 + $encoding.CodePage), $encoding) > $null;
+        }
+    }
+
+    # Get the first N bytes from the file, where N is the max preamble length we saw
+    [Byte[]]$bom = Get-Content -Encoding Byte -ReadCount $max_preamble_len -TotalCount $max_preamble_len -LiteralPath $path;
+
+    # Iterate through the sorted encodings, looking for a full match.
+    $found = $false;
+    Foreach ($encoding in $sortedlist.GetValueList()) {
+        $preamble = $encoding.GetPreamble();
+        If ($preamble -and $bom) {
+            Foreach ($i in 0..($preamble.Length - 1)) {
+                If ($i -ge $bom.Length) {
+                    break;
+                }
+                If ($preamble[$i] -ne $bom[$i]) {
+                    break;
+                }
+                ElseIf ($i + 1 -eq $preamble.Length) {
+                    $encodingObject = $encoding;
+                    $found = $true;
+                }
+            }
+            If ($found) {
+                break;
+            }
+        }
+    }
+    return $encodingObject;
+}
+
+function BackupFile {
+    param(
+        [String]$path,
+        [bool]$check_mode
+    );
+
+    $backuppath = $path + "." + [DateTime]::Now.ToString("yyyyMMdd-HHmmss");
+    Try {
+        Copy-Item $path $backuppath -WhatIf:$check_mode;
+    }
+    Catch {
+        $module.FailJson("Cannot copy backup file! ($($_.Exception.Message))", $_);
+    }
+    If (-not $check_mode) {
+        Try {
+            Get-Acl -Path $path | Set-Acl -Path $backuppath
+        }
+        Catch {
+            $module.FailJson("Cannot copy ACL to backup file! ($($_.Exception.Message))", $_);
+        }
+    }
+    return $backuppath;
+}
+
+If (Test-Path -LiteralPath $path -PathType "container") {
+    $module.FailJson("Path $path is a directory", $_);
+}
+
+$linesep = "`r`n"
+If ($newline -eq "unix") {
+    $linesep = "`n";
+}
+
+If ($insertbefore -and $insertafter) {
+    $module.Warn("Both insertbefore and insertafter parameters found, ignoring `"insertafter=$insertafter`"")
+}
+
+If (-not $insertbefore -and -not $insertafter) {
+    $insertafter = "EOF";
+}
+
+$insre = $null;
+If ($insertafter -and $insertafter -ne "BOF" -and $insertafter -ne "EOF") {
+    $insre = New-Object Regex $insertafter, 'Compiled';
+}
+ElseIf ($insertbefore -and $insertbefore -ne "BOF" -and $insertbefore -ne "EOF") {
+    $insre = New-Object Regex $insertbefore, 'Compiled';
+}
```

Comment:
This code makes it appear that `BOF` and `EOF` are supported for both `insertbefore` and `insertafter`, but the documentation only mentions `EOF` for `insertafter` and `BOF` for `insertbefore`.  Probably want to align documentation with code.